### PR TITLE
Adds generics to allow the use of kafka decoders and a return value t…

### PR DESCRIPTION
**Pull request by mortent**
This adds generics so it is possible to use kafka decoders to transform from the byte array to java objects before the messages pass through the filters. It also provides a return value through the chain of filters.

@charliek

### CODE REVIEW

File: src/main/java/smartthings/konsumer/KafkaListener.java [hunk 1]
```diff
@@ -18,32 +21,38 @@
...
 
-import java.util.EventListener;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.concurrent.*;
 
-public class KafkaListener {
+public class KafkaListener<K, V, R> {
```
Comment (charliek) [inline src/main/java/smartthings/konsumer/KafkaListener.java:35]: Might be good to document what the generics are here are so you can reference quickly?

File: src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java [hunk 2]
```diff
@@ -10,37 +10,37 @@
...
 
-	public MessageFilterChain(final MessageProcessor processor, final MessageFilter... messageFilters) {
+	public MessageFilterChain(final MessageProcessor<K, V, R> processor, final List<MessageFilter<K, V, R>> messageFilters) {
 		this.processor = processor;
-		this.filters = Collections.unmodifiableList(new ArrayList<MessageFilter>(Arrays.asList(messageFilters)));
+		this.filters = Collections.unmodifiableList(new ArrayList<>(messageFilters)); //make defensive copy
 	}
 
-	public void handle(MessageAndMetadata<byte[], byte[]> originalMessageAndMetadata,
+	public R handle(MessageAndMetadata<K, V> originalMessageAndMetadata,
```
Comment (charliek) [inline src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java:24]: Out of curiosity what use case caused you to want this? I originally thought of this as kind of like a servlet filter style pattern, but I can see some value in this.  Always could use Void if you don't want R to have a value.

**Review by charliek — COMMENTED** [review]
These look like good changes.  After reading all the code in the review the generic values make a lot more sense than when I first started.  What are your thoughts on adding doc comments? Worth it?

File: src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java (hunk 2 above)
Comment (mortent) [reply src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java:24]: I was in doubt about adding this, but it is helpful in some situations. 

What I needed it for, is to be able to calculate error ratios for message processing. In many situations a lot of messages will be thrown away quickly and therefore not be processed at all. The baseline number for calculating the error ratios includes all messages, so even a high percentage of processed events failing will not make a big impact on the error ratios. This allows us to use a return value to signal to the filters if a message was actually processed and should be included in the ratio calculations, or was simply thrown away.

File: src/main/java/smartthings/konsumer/KafkaListener.java (hunk 1 above)
Comment (mortent) [reply src/main/java/smartthings/konsumer/KafkaListener.java:35]: Yeah, I'll add some docs on the generics and the key classes/methods.

**Comment by charliek** [comment]
+1

Looks good to me. Merge when you get the docs in that you think should be there.

**Comment by charliek** [comment]
Also let me know if you want me to do a release. Happy to do one whenever.

**Comment by mortent** [comment]
@charliek I think we should do a release right now before we merge this change. It unfortunately breaks the backwards compatibility in the API, so we should try to avoid updating the snapshot version with these changes.

**Pull request merged**